### PR TITLE
TT-4091 - OAS Caching Paths support

### PR DESCRIPTION
**Pull request by vaske**
This PR:
- Adds `disabled` flag to cache plugin
- Implements migration from old cache string array to new advanced cache config
- Implements OAS conversion for advanced cache config

### CODE REVIEW

File: apidef/migration_test.go [hunk 1]
```diff
@@ -205,3 +205,48 @@ func TestAPIDefinition_MigrateVersioning_StripPath(t *testing.T) {
...
+		}
+	}
+
+	check := func(t *testing.T, base APIDefinition) {
+		err := base.MigrateCachePlugin()
+		assert.NoError(t, err)
+
+		defaultVersionInfo := base.VersionData.Versions[base.VersionData.DefaultVersion]
+		path1 := defaultVersionInfo.ExtendedPaths.AdvanceCacheConfig[0]
+		path1Cache := defaultVersionInfo.ExtendedPaths.Cached[0]
```
Comment (furkansenharputlu) [inline apidef/migration_test.go:241]: Not just one field check. Instead, we should write an `expectedAdvCache` and compare it with actual.

File: apidef/migration.go [hunk 2]
```diff
@@ -78,3 +78,37 @@ func (a *APIDefinition) MigrateVersioning() (versions []APIDefinition, err error
...
 	return
 }
+
+func (a *APIDefinition) MigrateCachePlugin() (err error) {
+	if a.VersionDefinition.Enabled || len(a.VersionDefinition.Versions) != 0 {
+		return errors.New("not migratable - new versioning is enabled")
+	}
+
+	if a.VersionData.NotVersioned && len(a.VersionData.Versions) > 1 {
+		return errors.New("not migratable - if not versioned, there should be 1 version info in versions map")
```
Comment (furkansenharputlu) [inline apidef/migration.go:88]: This part is already implemented in versioning function. No need to call here again

File: apidef/migration.go [hunk 3]
```diff
@@ -78,3 +78,37 @@ func (a *APIDefinition) MigrateVersioning() (versions []APIDefinition, err error
...
+	}
+
+	for vName, vInfo := range a.VersionData.Versions {
+		var updated = false
+		if vInfo.UseExtendedPaths && len(vInfo.ExtendedPaths.Cached) > 0 {
+			var methods = []CacheMeta{}
```
Comment (furkansenharputlu) [inline apidef/migration.go:94]: `methods` can be named as `advCacheMetas`

File: apidef/migration.go [hunk 4]
```diff
@@ -78,3 +78,37 @@ func (a *APIDefinition) MigrateVersioning() (versions []APIDefinition, err error
...
+func (a *APIDefinition) MigrateCachePlugin() (err error) {
+	if a.VersionDefinition.Enabled || len(a.VersionDefinition.Versions) != 0 {
+		return errors.New("not migratable - new versioning is enabled")
+	}
+
+	if a.VersionData.NotVersioned && len(a.VersionData.Versions) > 1 {
+		return errors.New("not migratable - if not versioned, there should be 1 version info in versions map")
+	}
+
+	for vName, vInfo := range a.VersionData.Versions {
```
Comment (furkansenharputlu) [inline apidef/migration.go:91]: We don't need to iterate over versions. As we will call this function after migrating versioning, every definition will have just one vInfo like `a.VersionData.Versions[""]`. See https://github.com/TykTechnologies/tyk/pull/3805/files#diff-e1d9b55a26f9d6225d56d6f0161959217308e5ad4d6934e7d7df4595d9c2a130R104

File: apidef/migration.go [hunk 5]
```diff
@@ -78,3 +78,37 @@ func (a *APIDefinition) MigrateVersioning() (versions []APIDefinition, err error
...
+		var updated = false
+		if vInfo.UseExtendedPaths && len(vInfo.ExtendedPaths.Cached) > 0 {
+			var methods = []CacheMeta{}
+			for _, cache := range vInfo.ExtendedPaths.Cached {
+				newCache := CacheMeta{
+					Path:     cache,
```
Comment (furkansenharputlu) [inline apidef/migration.go:97]: I know in old cache there is no `method` but how will we write it in OAS format if method is empty? We need to find a solution here.

File: apidef/migration.go [hunk 6]
```diff
@@ -78,3 +78,37 @@ func (a *APIDefinition) MigrateVersioning() (versions []APIDefinition, err error
...
+				methods = append(methods, newCache)
+			}
+			vInfo.ExtendedPaths.AdvanceCacheConfig = methods
+			updated = true
+		}
+
+		if updated {
+			delete(a.VersionData.Versions, vName)
+			a.VersionData.Versions[vName] = vInfo
+		}
```
Comment (furkansenharputlu) [inline apidef/migration.go:109]: No need this part

File: apidef/migration.go (hunk 5 above)
Comment (furkansenharputlu) [reply apidef/migration.go:97]: As cache works only for the safe methods, `GET`,`HEAD`, `OPTIONS`. We need to add 3 advance cache for all.

https://github.com/TykTechnologies/tyk/blob/f603dec31b91ae9c7221c990b078d82aa04bfece/gateway/mw_redis_cache.go#L377-L379

File: apidef/oas/middleware.go [hunk 7]
```diff
@@ -429,3 +432,76 @@ func (a *Allowance) ExtractTo(endpointMeta *apidef.EndPointMeta) {
...
+	}
+}
+
+type CacheMiddleware struct {
+	Enabled            bool   `bson:"enabled" json:"enabled"`
+	CacheByRegex       string `bson:"cacheByRegex,omitempty" json:"cacheByRegex,omitempty"`
+	CacheResponseCodes []int  `bson:"cacheResponseCodes,omitempty" json:"cacheResponseCodes,omitempty"`
+}
+
+func (a *CacheMiddleware) Fill(cm apidef.CacheMeta) {
```
Comment (furkansenharputlu) [inline apidef/oas/middleware.go:497]: Where is this called?

File: apidef/oas/middleware.go [hunk 8]
```diff
@@ -429,3 +432,76 @@ func (a *Allowance) ExtractTo(endpointMeta *apidef.EndPointMeta) {
...
+	a.Enabled = !cm.Disabled
+	a.CacheByRegex = cm.CacheKeyRegex
+	a.CacheResponseCodes = cm.CacheOnlyResponseCodes
+}
+
+func (a *CacheMiddleware) ExtractTo(endpointMeta *apidef.CacheMeta) {
```
Comment (furkansenharputlu) [inline apidef/oas/middleware.go:503]: where is this called?

Comment (vaske) [reply apidef/oas/middleware.go:503]: This is not called but I made it due to similar semantic we do everywhere

File: apidef/oas/middleware.go (hunk 7 above)
Comment (vaske) [reply apidef/oas/middleware.go:497]: on line 452 in same file.

File: apidef/migration.go (hunk 5 above)
Comment (vaske) [reply apidef/migration.go:97]: from where we can get info about method?

Comment (furkansenharputlu) [reply apidef/migration.go]: As I said above, you don’t get info about method. We need to write it to 3 safe methods.

File: apidef/migration_test.go [hunk 9]
```diff
@@ -315,3 +315,79 @@ func TestAPIDefinition_MigrateEndpointMeta(t *testing.T) {
...
+			cacheItemHead,
+			cacheItemOptions,
+		}
+
+		// for every cache item we have to have 3 new AdvanceCacheConfig items
+		// since we do support caching only for safe methods GET, OPTIONS and HEAD
+		assert.Equal(t, len(defaultVersionInfo.ExtendedPaths.AdvanceCacheConfig), len(defaultVersionInfo.ExtendedPaths.Cached)*3)
+		assert.Equal(t, path1.Path, path1Cache)
+
+		assert.Equal(t, expectedCachedDefault, defaultVersionInfo.ExtendedPaths.Cached)
```
Comment (furkansenharputlu) [inline apidef/migration_test.go:384]: After migration the old one should be empty.

File: apidef/oas/middleware.go [hunk 10]
```diff
@@ -179,6 +179,7 @@ func (ps Paths) Fill(ep apidef.ExtendedPathsSet) {
 	ps.fillAllowance(ep.BlackList, block)
 	ps.fillAllowance(ep.Ignored, ignoreAuthentication)
 	ps.fillMockResponse(ep.MockResponse)
+	ps.fillCache(ep)
```
Comment (furkansenharputlu) [inline apidef/oas/middleware.go:182]: Can we make `ep.AdvanceCacheConfig` like others?

File: apidef/oas/middleware.go [hunk 11]
```diff
@@ -475,3 +478,76 @@ type Header struct {
 	Name  string `bson:"name" json:"name"`
 	Value string `bson:"value" json:"value"`
 }
+
+func (ps Paths) fillCache(ep apidef.ExtendedPathsSet) {
+	advanceCacheConfig := &ep.AdvanceCacheConfig
+
+	for _, advCache := range *advanceCacheConfig {
```
Comment (furkansenharputlu) [inline apidef/oas/middleware.go:485]: we can take `cacheMetas []CacheMeta` as param. So we can get rid of 483.
`for _, cm := range cacheMetas {`

File: apidef/oas/middleware.go [hunk 12]
```diff
@@ -475,3 +478,76 @@ type Header struct {
...
+		if ps[advCache.Path] != nil {
+			plugins := ps[advCache.Path].getMethod(advCache.Method)
+			plugins.fillCache(advCache)
+
+		}
+	}
+}
+
+func (p *Plugins) fillCache(cacheMeta apidef.CacheMeta) {
```
Comment (furkansenharputlu) [inline apidef/oas/middleware.go:494]: We can get rid of this one if we apply above logic. That approach is used in other plugins too, it is good for consistency.

File: apidef/oas/middleware.go [hunk 13]
```diff
@@ -475,3 +478,76 @@ type Header struct {
...
+		newCache.CacheKeyRegex = p.Cache.CacheByRegex
+		newCache.Method = method
+		newCache.Path = path
+
+		advanceCacheConfig = append(advanceCacheConfig, newCache)
+		ep.AdvanceCacheConfig = advanceCacheConfig
+		return
+	}
+
+	founded := false
```
Comment (furkansenharputlu) [inline apidef/oas/middleware.go:517]: Why did we implement this part?

Comment (furkansenharputlu) [reply apidef/oas/middleware.go:517]: Can we do the same with below ? Let's keep consistency.
```
func (p *Plugins) extractMockResponseTo(ep *apidef.ExtendedPathsSet, path string, method string) {
	if p.MockResponse == nil {
		return
	}

	mockMeta := apidef.MockResponseMeta{Path: path, Method: method}
	p.MockResponse.ExtractTo(&mockMeta)
	ep.MockResponse = append(ep.MockResponse, mockMeta)
}
```

File: apidef/oas/middleware.go [hunk 14]
```diff
@@ -475,3 +478,76 @@ type Header struct {
...
+		newCache.CacheKeyRegex = p.Cache.CacheByRegex
+		newCache.Method = method
+		newCache.Path = path
+
+		advanceCacheConfig = append(advanceCacheConfig, newCache)
+		ep.AdvanceCacheConfig = advanceCacheConfig
+	}
+}
+
+type CacheMiddleware struct {
```
Comment (furkansenharputlu) [inline apidef/oas/middleware.go:537]: Let's rename it as `CachePlugin`, cache middleware is the one in the global level.

File: gateway/mw_version_check_test.go [hunk 15]
```diff
@@ -463,3 +464,60 @@ func TestOldVersioning_Expires(t *testing.T) {
 		})
 	})
 }
+
+func TestOld_Cache(t *testing.T) {
```
Comment (furkansenharputlu) [inline gateway/mw_version_check_test.go:468]: `TestOldCache`

File: gateway/mw_version_check_test.go [hunk 16]
```diff
@@ -463,3 +464,60 @@ func TestOldVersioning_Expires(t *testing.T) {
...
+
+	ts.Gw.LoadAPI(api())
+	_, _ = ts.Run(t, test.TestCase{Path: "/test", BodyMatch: "default", Code: http.StatusOK})
+
+	response = "updated"
+
+	check(t, api(), []test.TestCase{
+		{Path: "/test", BodyMatch: "default", Code: http.StatusOK},
+		{Path: "/anything", BodyMatch: "updated", Code: http.StatusOK},
+	})
```
Comment (furkansenharputlu) [inline gateway/mw_version_check_test.go:522]: This part should be `check` and we need to extract `t.Run("migration"` outside and call `check` inside migration again.

File: apidef/migration_test.go [hunk 17]
```diff
@@ -315,3 +315,79 @@ func TestAPIDefinition_MigrateEndpointMeta(t *testing.T) {
...
+				NotVersioned:   false,
+				DefaultVersion: name,
+				Versions: map[string]VersionInfo{
+					name: versionInfo,
+				},
+			},
+		}
+	}
+
+	check := func(t *testing.T, base APIDefinition) {
```
Comment (furkansenharputlu) [inline apidef/migration_test.go:345]: separate check function is not necessary here because you call it once

File: apidef/migration_test.go [hunk 18]
```diff
@@ -315,3 +315,79 @@ func TestAPIDefinition_MigrateEndpointMeta(t *testing.T) {
...
+		// for every cache item we have to have 3 new AdvanceCacheConfig items
+		// since we do support caching only for safe methods GET, OPTIONS and HEAD
+		assert.Equal(t, len(defaultVersionInfo.ExtendedPaths.AdvanceCacheConfig), len(defaultVersionInfo.ExtendedPaths.Cached)*3)
+		assert.Equal(t, path1.Path, path1Cache)
+
+		assert.Equal(t, expectedCachedDefault, defaultVersionInfo.ExtendedPaths.Cached)
+		assert.Equal(t, expectedAdvCacheMethods, defaultVersionInfo.ExtendedPaths.AdvanceCacheConfig)
+	}
+
+	t.Run("check migration of old data to new", func(t *testing.T) {
```
Comment (furkansenharputlu) [inline apidef/migration_test.go:388]: why did we put it inside a sub-test? This is an unnecessary layer. If we'd test more than one thing we'd add but this layer is the main one.

File: apidef/migration_test.go [hunk 19]
```diff
@@ -315,3 +315,79 @@ func TestAPIDefinition_MigrateEndpointMeta(t *testing.T) {
...
+			Path:          "test",
+			CacheKeyRegex: "",
+			Method:        http.MethodHead,
+		}
+		cacheItemOptions := CacheMeta{
+			Disabled:      false,
+			Path:          "test",
+			CacheKeyRegex: "",
+			Method:        http.MethodOptions,
+		}
```
Comment (furkansenharputlu) [inline apidef/migration_test.go:372]: Let's get rid of redundancy:
```
cacheItemHead:=cacheItemGet
cacheItemHead.Method = http.MethodHead

cacheItemOptions:=cacheItemGet
cacheItemOptions.Method = http.MethodOptions
```

File: apidef/migration_test.go [hunk 20]
```diff
@@ -315,3 +315,79 @@ func TestAPIDefinition_MigrateEndpointMeta(t *testing.T) {
...
+		expectedAdvCacheMethods := []CacheMeta{
+			cacheItemGet,
+			cacheItemHead,
+			cacheItemOptions,
+		}
+
+		// for every cache item we have to have 3 new AdvanceCacheConfig items
+		// since we do support caching only for safe methods GET, OPTIONS and HEAD
+		assert.Equal(t, len(defaultVersionInfo.ExtendedPaths.AdvanceCacheConfig), len(defaultVersionInfo.ExtendedPaths.Cached)*3)
+		assert.Equal(t, path1.Path, path1Cache)
```
Comment (furkansenharputlu) [inline apidef/migration_test.go:382]: In the below, we already check, these can be removed I think
```
assert.Equal(t, expectedAdvCacheMethods, defaultVersionInfo.ExtendedPaths.AdvanceCacheConfig)
```

File: apidef/migration_test.go [hunk 21]
```diff
@@ -315,3 +315,79 @@ func TestAPIDefinition_MigrateEndpointMeta(t *testing.T) {
 	assert.Equal(t, expectedWhitelist, api.VersionData.Versions[""].ExtendedPaths.Ignored)
 	assert.Equal(t, expectedMockResponse, api.VersionData.Versions[""].ExtendedPaths.MockResponse)
 }
+
+func TestAPIDefinition_MigrateCache(t *testing.T) {
+	name := ""
```
Comment (furkansenharputlu) [inline apidef/migration_test.go:320]: ?

File: apidef/migration.go [hunk 22]
```diff
@@ -188,3 +189,34 @@ func (a *APIDefinition) Migrate() (versions []APIDefinition, err error) {
...
+				Disabled: false,
+				Method:   http.MethodHead,
+			}
+			newOptionsMethodCache := CacheMeta{
+				Path:     cache,
+				Disabled: false,
+				Method:   http.MethodOptions,
+			}
+			advCacheMethods = append(advCacheMethods, newGetMethodCache, newHeadMethodCache, newOptionsMethodCache)
+		}
```
Comment (furkansenharputlu) [inline apidef/migration.go:216]: let's skip a line after `for`

File: apidef/migration.go [hunk 23]
```diff
@@ -188,3 +189,34 @@ func (a *APIDefinition) Migrate() (versions []APIDefinition, err error) {
...
+		vInfo.ExtendedPaths.AdvanceCacheConfig = advCacheMethods
+	}
```
Comment (furkansenharputlu) [inline apidef/migration.go:218]: let's skip a line after `if`

File: apidef/migration_test.go [hunk 24]
```diff
@@ -315,3 +315,79 @@ func TestAPIDefinition_MigrateEndpointMeta(t *testing.T) {
...
+
+		assert.Equal(t, expectedCachedDefault, defaultVersionInfo.ExtendedPaths.Cached)
+		assert.Equal(t, expectedAdvCacheMethods, defaultVersionInfo.ExtendedPaths.AdvanceCacheConfig)
+	}
+
+	t.Run("check migration of old data to new", func(t *testing.T) {
+		base := old()
+		check(t, base)
+	})
+
```
Comment (furkansenharputlu) [inline apidef/migration_test.go:392]: unnecessary empty line

File: apidef/migration.go [hunk 25]
```diff
@@ -188,3 +189,34 @@ func (a *APIDefinition) Migrate() (versions []APIDefinition, err error) {
...
+func (a *APIDefinition) MigrateCachePlugin() (err error) {
+	vInfo := a.VersionData.Versions[""]
+	list := vInfo.ExtendedPaths.Cached
+
+	if vInfo.UseExtendedPaths && len(list) > 0 {
+		var advCacheMethods = []CacheMeta{}
+		for _, cache := range list {
+			newGetMethodCache := CacheMeta{
+				Path:     cache,
+				Disabled: false,
```
Comment (furkansenharputlu) [inline apidef/migration.go:202]: `Disabled`s are already `false`. Why would we add?

File: apidef/oas/middleware.go [hunk 26]
```diff
@@ -475,3 +478,76 @@ type Header struct {
...
+	CacheResponseCodes []int  `bson:"cacheResponseCodes,omitempty" json:"cacheResponseCodes,omitempty"`
+}
+
+func (a *CacheMiddleware) Fill(cm apidef.CacheMeta) {
+	a.Enabled = !cm.Disabled
+	a.CacheByRegex = cm.CacheKeyRegex
+	a.CacheResponseCodes = cm.CacheOnlyResponseCodes
+}
+
+func (a *CacheMiddleware) ExtractTo(endpointMeta *apidef.CacheMeta) {
```
Comment (furkansenharputlu) [inline apidef/oas/middleware.go:549]: `endpointMeta` -> `cm`

File: apidef/migration_test.go (hunk 20 above)
Comment (vaske) [reply apidef/migration_test.go:382]: yeah but I applied here logic the more the merrier :)

File: apidef/migration_test.go (hunk 21 above)
Comment (vaske) [reply apidef/migration_test.go:320]: leftover will put inside then exposed here

File: apidef/migration.go (hunk 23 above)
Comment (vaske) [reply apidef/migration.go:218]: is that by some convention?

File: gateway/mw_version_check_test.go [hunk 27]
```diff
@@ -463,3 +464,57 @@ func TestOldVersioning_Expires(t *testing.T) {
...
+	check := func(t *testing.T, response string) {
+
+		_, _ = ts.Run(t, []test.TestCase{
+			{Path: "/test", BodyMatch: "default", Code: http.StatusOK},
+			{Path: "/anything", BodyMatch: response, Code: http.StatusOK},
+		}...)
+
+	}
+
+	ts.Gw.LoadAPI(api())
```
Comment (furkansenharputlu) [inline gateway/mw_version_check_test.go:510]: I think load should be done inside check function. Right now,  migrated api is not loaded. Check function should get api param.

File: apidef/oas/middleware.go [hunk 28]
```diff
@@ -521,3 +524,55 @@ func (mt *MethodTransform) ExtractTo(meta *apidef.MethodTransformMeta) {
...
+}
+
+func (p *Plugins) extractCacheTo(ep *apidef.ExtendedPathsSet, path string, method string) {
+
+	if p.Cache == nil {
+		return
+	}
+
+	newCacheMeta := apidef.CacheMeta{
+		CacheOnlyResponseCodes: p.Cache.CacheResponseCodes,
```
Comment (furkansenharputlu) [inline apidef/oas/middleware.go:552]: ExtractTo function already fills this fields. Just method and path should be defined outside.

**Review by furkansenharputlu — APPROVED** [review]
Added 2 small comments. Overall LGTM!

**Review by jeffy-mathew — APPROVED** [review]

**Pull request merged**